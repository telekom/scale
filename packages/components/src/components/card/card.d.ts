import { HTMLStencilElement } from '../../../dist/types/stencil.core';
export declare class Card {
    hostElement: HTMLStencilElement;
    size?: string;
    theme?: string;
    variant?: string;
    disabled?: boolean;
    deselected?: boolean;
    imageTop?: string;
    imageTopAlt?: string;
    private hasSlotHeader;
    private hasSlotFooter;
    private getCssClassMap;
    componentWillLoad(): void;
    render(): any;
}
