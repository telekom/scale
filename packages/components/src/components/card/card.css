.card {
	box-sizing: border-box;
	background: var(--card-background-color, #fff);
	border-radius: var(--card-border-radius, 4px); 
	width: var(--card-width, 400px);
	max-width: var(--card-max-width, 550px);
	box-shadow: var(--card-box-shadow, 0 2px 4px 0 rgba(0, 0, 0, 0.12));
	border: var(--card-border-color, 1px solid rgba(0, 0, 0, 0.1)); 
}

.card__body {
	padding: var(--card-body-padding, 1rem); 
}

.card__header {
	padding: var(--card-header-padding);
}

.card__footer {
	padding: var(--card-footer-padding);
}