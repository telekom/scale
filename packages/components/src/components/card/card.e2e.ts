import { newE2EPage } from '@stencil/core/testing';

describe('t-card', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<t-card>Click me!</t-card>');
    const element = await page.find('t-card');
    expect(element).toHaveClass('hydrated');
  });
});
// Todo: Add test for image
// Todo: Add test for image alt text