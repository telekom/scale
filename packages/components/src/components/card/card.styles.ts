import { JssStyle } from 'jss';
export const styles: JssStyle = {
  card: {
    boxSizing: 'border-box',
    overflow: 'hidden',
    color: '#333',
    background: '#fff',
    border: '1px solid rgba(0, 0, 0, 0.1)',
    borderRadius: 4,
    width: '100%',
    maxWidth: 400,
    boxShadow: '0 2px 4px 0 rgba(0, 0, 0, 0.12)',
    transition: 'all .2s ease-in-out',
  },
  card__body: {
    padding: '1rem',
    '& ::slotted(*)': {
      margin: 0,
    },
  },
  'card--interactive': {
    textDecoration: 'none',
    cursor: 'pointer',
    display: 'block',
    '&:hover': {
      boxShadow: '0 2px 8px 0 rgba(0, 0, 0, 0.24)',
    },
    '&:focus': {
      boxShadow: '0 2px 8px 0 rgba(0, 0, 0, 0.24)',
    },
    '&:active': {
      boxShadow: '0 2px 8px 0 rgba(0, 0, 0, 0.24)',
    },
    '& $card__body ::slotted(*)': {
      margin: 0,
    },
  },
  'card--disabled': {
    '&, &:hover, &:focus, &:active': {
      opacity: 0.5,
      cursor: 'not-allowed',
      boxShadow: '0 2px 4px 0 rgba(0, 0, 0, 0.12)',
      outline: 'none',
    },
  },
};
