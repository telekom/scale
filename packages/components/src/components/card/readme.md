# scale-card

<!-- Auto Generated Below -->


## Usage

### Card

# card

<scale-card>
  <h1>Card Title</h1>
  <p>This is card content</p>
</scale-card>

## interactive card

<scale-card href="http://example.com" target="_blank">
  <p>click me</p>
</scale-card>



## Properties

| Property      | Attribute      | Description                    | Type                                     | Default     |
| ------------- | -------------- | ------------------------------ | ---------------------------------------- | ----------- |
| `customClass` | `custom-class` | (optional) Card class          | `string`                                 | `''`        |
| `disabled`    | `disabled`     | (optional) Link disabled       | `boolean`                                | `false`     |
| `href`        | `href`         | (optional) Link card           | `string`                                 | `''`        |
| `interactive` | `interactive`  | (optional) Link interactive    | `boolean`                                | `false`     |
| `styles`      | --             | (optional) Injected jss styles | `StyleSheet<string \| number \| symbol>` | `undefined` |
| `target`      | `target`       | (optional) Link card target    | `string`                                 | `'_self'`   |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
