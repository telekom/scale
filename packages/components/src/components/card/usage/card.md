# Default

<scale-card>
  <h1>Card Title</h1>
  <p>This is card content</p>
</scale-card>
