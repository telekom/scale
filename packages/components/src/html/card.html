<!DOCTYPE html>
<html dir="ltr" lang="en">
  <head>
    <meta charset="utf-8" />
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0"
    />
    <title>Stencil Component Starter</title>

    <script type="module" src="/build/scale-components.esm.js"></script>
    <link rel="stylesheet" href="/build/scale-components.css" />
    <style type="text/css" media="screen, print">
      html {
        margin: 0;
        padding: 0;
      }

      body {
        margin: 0;
        padding: 4rem;
      }

      section {
        display: inline-grid;
        width: 18vw;
      }
    </style>
  </head>

  <body>
    <h1>card</h1>

    <section>
      <h3>icon</h3>
      <scale-card label="Icon" id="icon-card" to="javascript:void(0);">
        <div style="display: flex; justify-content: space-between">
          <span style="font: var(--telekom-text-style-heading-5)">Icon</span>
          <scale-menu-flyout>
            <scale-button
              slot="trigger"
              variant="secondary"
              icon-only
              size="small"
              id="flyout-trigger"
            >
              <scale-icon-service-settings
                accessibility-title="Open menu"
              ></scale-icon-service-settings>
            </scale-button>
            <scale-menu-flyout-list>
              <scale-menu-flyout-item>Menu Item 1</scale-menu-flyout-item>
              <scale-menu-flyout-item>Menu Item 2</scale-menu-flyout-item>
              <scale-menu-flyout-item>Menu Item 3</scale-menu-flyout-item>
            </scale-menu-flyout-list>
          </scale-menu-flyout>
        </div>
        <div
          style="
            display: flex;
            flex-direction: row;
            justify-content: center;
            align-items: center;
            border-radius: var(--telekom-radius-small);
            background: var(--telekom-color-functional-informational-subtle);
            color: var(--telekom-color-text-and-icon-link-standard);
            margin-top: var(--telekom-spacing-unit-x6);
            height: 178px;
          "
        >
          Content area
        </div>
      </scale-card>

      <script>
        document.getElementById('icon-card').addEventListener('click', (e) => {
          if (!e.path.find(({ id }) => id === 'flyout-trigger')) {
            e.stopImmediatePropagation();

            document.getElementById('flyout-trigger').dispatchEvent(
              new MouseEvent('click', {
                bubbles: true,
                cancelable: false,
              })
            );
          }
        });
      </script>

      <h3>checkbox</h3>

      <scale-card label="Checkbox" id="checkbox-card" to="javascript:void(0);">
        <div
          style="
            display: flex;
            justify-content: space-between;
            align-items: center;
          "
        >
          <span style="font: var(--telekom-text-style-heading-5)"
            >Checkbox</span
          >
          <scale-checkbox id="checkbox" checked></scale-checkbox>
        </div>
        <div
          style="
            display: flex;
            flex-direction: row;
            justify-content: center;
            align-items: center;
            border-radius: var(--telekom-radius-small);
            background: var(--telekom-color-functional-informational-subtle);
            color: var(--telekom-color-text-and-icon-link-standard);
            margin-top: var(--telekom-spacing-unit-x6);
            width: 100%;
            height: 178px;
          "
        >
          Content area
        </div>
      </scale-card>
      <script>
        const checkbox = document.getElementById('checkbox');
        document
          .getElementById('checkbox-card')
          .addEventListener('click', () => {
            if (!checkbox.hasAttribute('checked')) {
              checkbox.setAttribute('checked', 'true');
            } else {
              checkbox.removeAttribute('checked');
            }
          });
      </script>
    </section>

    <section>
      <h2>Movable</h2>

      <h3>one</h3>
      <scale-card
        movable="true"
        to="https://example.com"
        target="_blank"
        rel="noopener noreferrer"
        label="Example Card"
        >A card with a link to example.com</scale-card
      >
      <h3>two</h3>

      <scale-card movable="true" to="https://example.com" target="_blank">
        <div style="margin: calc(-1 * var(--telekom-spacing-unit-x4))">
          <img
            style="margin-bottom: -4px; width: 100%"
            src="assets/card-grayscale.jpeg"
            alt="placeholder"
          />
          <div style="padding: var(--telekom-spacing-unit-x4)">
            <div style="font: var(--telekom-text-style-heading-5)">
              Powered by AI
            </div>
            <div
              style="
                font: var(--telekom-text-style-body);
                color: var(--telekom-color-text-and-icon-additional);
              "
            >
              Learn about current use cases for artifical intelligence.
            </div>
          </div>
        </div>
      </scale-card>

      <h3>three</h3>

      <div style="max-width: 200px">
        <scale-card movable="true" label="Example Card with further functions">
          <article style="margin: -24px">
            <img
              style="margin-bottom: -4px; width: 100%"
              src="https://placeimg.com/200/200/tech/grayscale"
              alt="placeholder"
            />
            <div style="padding: 24px">
              <div style="padding: 0 0 24px 0">
                A card with with further functions
              </div>
              <div
                style="
                  padding: 0 0 24px 0;
                  display: flex;
                  justify-content: center;
                "
              >
                <scale-button>Label</scale-button>
              </div>
              <div style="display: flex; justify-content: flex-end">
                <scale-link href="https://example.com" tabindex="0"
                  >More...</scale-link
                >
              </div>
            </div>
          </article>
        </scale-card>
      </div>
    </section>
  </body>
</html>
