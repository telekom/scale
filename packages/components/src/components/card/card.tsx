import { Component, Prop, h, Host } from '@stencil/core';
import { CssClassMap } from '../../utils/utils';
import classNames from 'classnames';
import { styles } from './card.styles';
import { CssInJs } from '../../utils/css-in-js';
import { StyleSheet } from 'jss';
import Base from '../../utils/base-interface';

@Component({
  tag: 'scale-card',
  shadow: true,
})
export class Card implements Base {
  /** (optional) Card class */
  @Prop() customClass?: string = '';
  /** (optional) Link card */
  @Prop() href?: string = '';
  /** (optional) Link card target */
  @Prop() target?: string = '_self';
  /** (optional) Link interactive */
  @Prop() interactive?: boolean = false;
  /** (optional) Link disabled */
  @Prop() disabled?: boolean = false;
  /** (optional) Injected jss styles */
  @Prop() styles?: StyleSheet;
  /** decorator Jss stylesheet */
  @CssInJs('Card', styles) stylesheet: StyleSheet;

  componentWillLoad() {}
  componentWillUpdate() {}

  render() {
    const { classes } = this.stylesheet;
    const Tag = !!this.href ? 'a' : 'div';

    return (
      <Host>
        <style>{this.stylesheet.toString()}</style>
        <Tag
          class={this.getCssClassMap()}
          {...(!!this.href ? { href: this.href } : {})}
          {...(!!this.href ? { target: this.target } : {})}
          {...(!!this.interactive ? { tabindex: 1 } : {})}
        >
          <div class={classes.card__body}>
            <slot />
          </div>
        </Tag>
      </Host>
    );
  }

  getCssClassMap(): CssClassMap {
    const { classes } = this.stylesheet;

    return classNames(
      classes.card,
      this.customClass && this.customClass,
      (!!this.href || this.interactive) && classes[`card--interactive`],
      this.disabled && classes['card--disabled']
    );
  }
}
