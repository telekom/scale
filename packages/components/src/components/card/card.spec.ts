import { newSpecPage } from '@stencil/core/testing';
import { Card } from './card';

describe('Card', () => {
	let element;
	beforeEach(async () => {
		element = new Card()
	});

	it('should match snapshot', async () => {
		const page = await newSpecPage({
			components: [Card],
			html: `<t-card>Label</t-card>`,
		});
		expect(page.root).toMatchSnapshot();
    });

    it('should match snapshot with image', async () => {
		const page = await newSpecPage({
			components: [Card],
			html: `<t-card image-top="http://placehold.it/400x300">A title</t-card>`,
		});
		expect(page.root).toMatchSnapshot();
    });

	it('should have a default css class', () => {
		expect(element.getCssClassMap()).toBe('card');
	})

    it('should handle size css class', () => {
		element.size = 'small'
		expect(element.getCssClassMap()).toContain('card--size-small');
    })
    
    it('should handle theme css class', () => {
		element.theme = 'default';
		expect(element.getCssClassMap()).toContain('card--theme-default');
    })
    
    it('should handle variant css class', () => {
		element.variant = 'primary';
		expect(element.getCssClassMap()).toContain('card--variant-primary');
    })
    
    it('should handle disabled css class', () => {
        element.disabled = true;
		expect(element.getCssClassMap()).toContain('card--disabled');
    })
    
    it('should handle deselected css class', () => {
		element.deselected = true;
		expect(element.getCssClassMap()).toContain('card--deselected');
	})
});

