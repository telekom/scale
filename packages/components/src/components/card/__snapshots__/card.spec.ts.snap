// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match snapshot 1`] = `
<scale-card>
  <mock:shadow-root>
    <style>
      .card-0-0-5 {
  color: #333;
  width: 100%;
  border: 1px solid rgba(0, 0, 0, 0.1);
  overflow: hidden;
  max-width: 400px;
  background: #fff;
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.12);
  box-sizing: border-box;
  transition: all .2s ease-in-out;
  border-radius: 4px;
}
.card__body-0-0-6 {
  padding: 1rem;
}
.card__body-0-0-6 ::slotted(*) {
  margin: 0;
}
.card--interactive-0-0-7 {
  cursor: pointer;
  display: block;
  text-decoration: none;
}
.card--interactive-0-0-7:hover {
  box-shadow: 0 2px 8px 0 rgba(0, 0, 0, 0.24);
}
.card--interactive-0-0-7:focus {
  box-shadow: 0 2px 8px 0 rgba(0, 0, 0, 0.24);
}
.card--interactive-0-0-7:active {
  box-shadow: 0 2px 8px 0 rgba(0, 0, 0, 0.24);
}
.card--interactive-0-0-7 .card__body-0-0-6 ::slotted(*) {
  margin: 0;
}
.card--disabled-0-0-8 {}
.card--disabled-0-0-8, .card--disabled-0-0-8:hover, .card--disabled-0-0-8:focus, .card--disabled-0-0-8:active {
  cursor: not-allowed;
  opacity: 0.5;
  outline: none;
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.12);
}
    </style>
    <div class="card-0-0-5">
      <div class="card__body-0-0-6">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
  Label
</scale-card>
`;
